azurerm/resource_arm_sql_database.go

```diff
@@ -118,128 +119,157 @@ func resourceArmSqlDatabase() *schema.Resource {
 	}
 }
 
-func resourceArmSqlDatabaseCreate(d *schema.ResourceData, meta interface{}) error {
-	client := meta.(*ArmClient)
-	rivieraClient := client.rivieraClient
+func resourceArmSqlDatabaseCreateOrUpdate(d *schema.ResourceData, meta interface{}) error {
+	sqlDatabasesClient := meta.(*ArmClient).sqlDatabasesClient
 
-	tags := d.Get("tags").(map[string]interface{})
-	expandedTags := expandTags(tags)
-
-	command := &sql.CreateOrUpdateDatabase{
-		Name:              d.Get("name").(string),
-		Location:          d.Get("location").(string),
-		ResourceGroupName: d.Get("resource_group_name").(string),
-		ServerName:        d.Get("server_name").(string),
-		Tags:              *expandedTags,
-		CreateMode:        azure.String(d.Get("create_mode").(string)),
-	}
+	databaseName := d.Get("name").(string)
+	resGroup := d.Get("resource_group_name").(string)
+	location := d.Get("location").(string)
+	serverName := d.Get("server_name").(string)
+	sourceDatabaseID := d.Get("source_database_id").(string)
+	collation := d.Get("collation").(string)
+	elasticPoolName := d.Get("elastic_pool_name").(string)
 
-	if v, ok := d.GetOk("source_database_id"); ok {
-		command.SourceDatabaseID = azure.String(v.(string))
-	}
+	createMode := sql.CreateMode(d.Get("create_mode").(string))
 
-	if v, ok := d.GetOk("edition"); ok {
-		command.Edition = azure.String(v.(string))
-	}
+	maxSizeBytes := d.Get("max_size_bytes").(string)
 
-	if v, ok := d.GetOk("collation"); ok {
-		command.Collation = azure.String(v.(string))
-	}
+	edition := sql.DatabaseEdition(d.Get("edition").(string))
 
-	if v, ok := d.GetOk("max_size_bytes"); ok {
-		command.MaxSizeBytes = azure.String(v.(string))
-	}
+	requestedServiceObjectiveIDString := d.Get("requested_service_objective_id").(string)
+	var requestedServiceObjectiveID uuid.UUID
 
-	if v, ok := d.GetOk("source_database_deletion_date"); ok {
-		command.SourceDatabaseDeletionDate = azure.String(v.(string))
-	}
+	requestedServiceObjectiveNameString := d.Get("requested_service_objective_name").(string)
+	var requestedServiceObjectiveName sql.ServiceObjectiveName
+
+	if requestedServiceObjectiveIDString != "" && requestedServiceObjectiveNameString == "" {
```

Comment:
we can use the `conflicts_with` option in the Schema to handle this for us automatically, i.e.:

```
"requested_service_objective_id": {
    Type:     schema.TypeString,
    Optional: true,
    Computed: true,
    ConflictsWith: []string{ "requested_service_objective_name" },
},
 
"requested_service_objective_name": {
    Type:     schema.TypeString,
    Optional: true,
    Computed: true,
    ConflictsWith: []string{ "requested_service_objective_id" },
},
```

and then just have if statements for each field, and assign the value directly to props as needed, rather than checking the state of both:
```
if v, ok := d.Get("requested_service_objective_id").(string); ok {
  rsoid, err := uuid.FromString(v.(string))
   if err != nil {
     return err
  }
  props.RequestedServiceObjectiveID  = &rsoid
}

if v, ok := d.Get("requested_service_objective_name").(string); ok {
  props.RequestedServiceObjectiveID  = sql.ServiceObjectiveName(v.(string))
}
```